cmd/swarm/swarm-smoke/pss.go

```diff
@@ -0,0 +1,237 @@
+			continue
+		}
+
+		// subscribe to the topic for the session
+		// this creates the incoming message handler automatically
+		// any message received with this topic comes on the channel
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+		defer cancel()
+		sub, err := rpcClient.Subscribe(ctx, "pss", session.msgC, "receive", session.topic, true, false)
+		if err != nil {
+			log.Error("Error calling host for subscribe", "err", err)
+			continue
+		}
+		session.nodes = append(session.nodes,
+			&pssNode{
+				hostIdx:        i,
+				addr:           []byte(addr),
+				client:         rpcClient,
+				pubkey:         pubkey,
+				deregisterFunc: sub.Unsubscribe,
+			},
+		)
+	}
+
+	return session
+}
+
+// sends a single PSS message between two random nodes using asymetric encryption
+func pssAsymDo(ctx *cli.Context, session *pssSession, tuid string) error {
+
+	senderNodeIdx := rand.Intn(len(session.nodes) - 1)
+	senderNode := session.nodes[senderNodeIdx]
+	log.Trace("sender node", "pss_baseAddr", hexutil.Encode(senderNode.addr), "host", hosts[senderNodeIdx])
+
+	// receiving node has to be different than sender
+	var recvNodeIdx int
+	for i := 0; i < 20; i++ {
+		recvNodeIdx = rand.Intn(len(session.nodes) - 1)
+		if recvNodeIdx != senderNodeIdx {
+			break
+		}
+	}
+	if recvNodeIdx == senderNodeIdx {
+		return fmt.Errorf("sender and receiving nodes can't be the same. Do you have enough hosts?")
+	}
+
+	recvNode := session.nodes[recvNodeIdx]
+	log.Trace("recv node", "pss_baseAddr", hexutil.Encode(recvNode.addr), "host", hosts[recvNodeIdx])
+
+	// set recipient in pivot node
+	err := senderNode.client.Call(nil, "pss_setPeerPublicKey", recvNode.pubkey, session.topic, "0x")
+	if err != nil {
+		return err
+	}
+
+	// create new message and add it to job index to check for receives
+	randomMsg := testutil.RandomBytes(seed, 128)
+	// change seed so that the next random message is different
+	seed = seed + 1
+
+	msgIdx := toMsgIdx(randomMsg)
+	session.requiredJobs[msgIdx] = &pssJob{
+		msg:  randomMsg,
+		mode: pssModeAsym,
+	}
+
+	// send the message
+	hostIdx := session.nodes[recvNodeIdx].hostIdx
+	log.Info("sending msg", "job", hexutil.Encode([]byte(msgIdx)), "sender", hosts[senderNodeIdx], "recv", hosts[hostIdx])
+	err = senderNode.client.Call(nil, "pss_sendAsym", recvNode.pubkey, session.topic, hexutil.Encode(randomMsg))
+	if err != nil {
+		return err
+	}
+
+	// receive the message and check for its content
+	pctx, cancel := context.WithTimeout(context.TODO(), time.Second)
+	defer cancel()
+	select {
+	case res := <-(session.msgC):
+		resIdx := toMsgIdx(res.Msg)
+		resJob, ok := session.requiredJobs[resIdx]
+		if !ok {
+			return fmt.Errorf("corrupt message for job %s", resIdx)
+		}
+		if !bytes.Equal(res.Msg, resJob.msg) {
+			return fmt.Errorf("message mismatch. expected: %s got: %s", resJob.msg, res.Msg)
+		}
+		log.Info("got msg", "job", hexutil.Encode([]byte(resIdx)))
+	case <-pctx.Done():
```

Comment:
you can just use `time.After(1*time.Second)` - it is simpler than context with a timeout and cancel function.